### PR TITLE
deprecate zero-arg Matrix constructors (#14201)

### PR DESCRIPTION
**Pull request by StefanKarpinski**

### CODE REVIEW

**Comment by JeffBezanson** [comment]
Need to pick between this and #20250. I vote for deprecating the zero-arg constructors; `Matrix()` does not seem very meaningful to me.

File: test/arrayops.jl [hunk 1]
```diff
@@ -178,7 +178,6 @@ end
     @test convert(Array{Int}, a) == a
     @test convert(Array{Float64}, a) == a
     @test convert(Matrix, a) == a
-    @test convert(Matrix{Int}, a) == a
```
Comment (tkelman) [inline test/arrayops.jl]: why is this removed? still seems valid to me

File: test/arrayops.jl [hunk 2]
```diff
@@ -271,18 +270,19 @@ end
...
     @test size(Vector{Int}()) == (0,)
     @test size(Vector(3)) == (3,)
     @test size(Vector()) == (0,)
     @test size(Matrix{Int}(2,3)) == (2,3)
-    @test size(Matrix{Int}()) == (0,0)
     @test size(Matrix(2,3)) == (2,3)
-    @test size(Matrix()) == (0,0)
+
+    # TODO: will throw MethodError after 0.6 deprecations are deleted
+    @test_throws ErrorException Matrix{Int}()
```
Comment (tkelman) [inline test/arrayops.jl]: it's not a good idea to test depwarns like this, --depwarn=error doesn't always apply, for example if running with JULIA_CPU_CORES=1

Comment (StefanKarpinski) [inline test/arrayops.jl]: Then how do you recommend doing it?

File: test/arrayops.jl (hunk 1 above)
Comment (StefanKarpinski) [inline test/arrayops.jl]: accident

File: test/arrayops.jl (hunk 2 above)
Comment (tkelman) [inline test/arrayops.jl]: probably leave it out for now unless we want to add a depwarn_exec.jl file that we run in a separate process with known --depwarn settings.

if `JULIA_CPU_CORES=1 make test-arrayops` passes on this branch then this might be fine

File: test/arrayops.jl [hunk 3]
```diff
@@ -271,18 +271,30 @@ end
...
+    elseif dw == 1
+        @test_warn "deprecated" Matrix{Int}()
+        @test_warn "deprecated" Matrix()
+    elseif dw == 0
+        @test size(Matrix{Int}()) == (0,0)
+        @test size(Matrix()) == (0,0)
+    else
+        error("unexpected depwarn value")
+    end
+    @test_throws MethodError Array{Int,3}()
```
Comment (StefanKarpinski) [inline test/arrayops.jl]: We could have a `@test_deprecated` macro that does this for you. Usage would be something like this:

```jl
@test_deprecated size(Matrix{Int}()) == (0,0)
```

which would expand to one of the following, depending on the value of `Base.JLOptions().depwarn`:

```jl
@test_throws ErrorException size(Matrix{Int}())
@test_warn "deprecated" size(Matrix{Int}())
@test size(Matrix{Int}()) == (0,0)
```

The main reason I wanted this complexity in here is so that I couldn't possibly forget to update the tests when we delete the deprecations for 1.0.

Comment (tkelman) [inline test/arrayops.jl]: that's a very good idea

Comment (tkelman) [inline test/arrayops.jl]: though we probably want to put such tests in a single place and make sure they get run regularly with all 3 values of the flag (or expand to 3 tests that modify the option in a scoped way, if that's possible inside julia)

File: base/sysimg.jl [hunk 4]
```diff
@@ -96,7 +96,6 @@ include("subarray.jl")
 (::Type{Vector})() = Array{Any,1}(0)
```
Comment (stevengj) [inline base/sysimg.jl]: I would prefer to deprecate `Vector()` as well.

Comment (StefanKarpinski) [inline base/sysimg.jl]: I would too but I have a suspicion it's fairly widely used, so I wanted to do the uncontroversial part first. It's fairly common to initialize a vector with zero elements and there was a period where `[]` created a Void eltype empty array. There's also a consistency argument here: if we deprecate that we should also deprecate `Vector{T}()` – are you in favor of that as well?

Comment (stevengj) [inline base/sysimg.jl]: Yes.

Comment (stevengj) [inline base/sysimg.jl]: I grepped for `Vector()` or `Vector{Foo}()` in all registered packages, and you're right that quite a few packages use it:

ASTInterpreter,AWS,AbstractTables,ApproxFun,Bio,BlackBoxOptim,Bukdu,CBOR,Celeste,Church,Clipper,Clustering,Compat,ControlCore,ControlSystems,Cxx,DSGE,DWARF,DataCubes,DataFrames,DeterministicPolicyGradient,DifferentialDynamicProgramming,Diversity,Dopri,EMIRT,FlatBuffers,FunctionalCollections,GUITestRunner,Gallium,Graft,Hecke,IndexableBitVectors,IntArrays,IntervalWavelets,Isotonic,JuliaFEM,LLVM,LightGraphs,LightGraphsExtras,Lumberjack,Mamba,MetadataTools,Metamath,MultidimensionalTables,MultipleTesting,Named,NearestNeighbors,NetworkFlows,NullableArrays,ODBC,ODEInterface,ObjFileBase,OpenDSSDirect,ParallelAccelerator,ParserCombinator,PlotRecipes,PlyIO,Primes,QuantumLab,RData,ResettableStacks,RigidBodyDynamics,RouletteWheels,SALSA,SemidefiniteModel,SemidefiniteModels,StaticArrays,StructuredQueries,SugarBLAS,Sundials,TensorDecompositions,TerminalUI,TestRunner,ThermodynamicsTable,TimeSeries,TimeSeriesIO,TravelingSalesmanHeuristics,Turing,TypedTables,Voronoi,WiltonInts84,WordNet

Comment (stevengj) [inline base/sysimg.jl]: A much smaller number of packages use `Matrix()` or `Matrix{T}()`:

Bio,DSGE,DifferentialDynamicProgramming,Elemental,JPLEphemeris,JuliaFEM,LightGraphs,Mamba,NearestNeighbors,OpenDSSDirect,ParallelAccelerator

Comment (martinholters) [inline base/sysimg.jl]: I'd also favor deprecating `Vector()` and `Vector{T}()`, but doing so in a separate PR due to the expected controversy sounds like a good idea.

Comment (yurivish) [reply base/sysimg.jl]: Now that we have femtocleaner would this be a less controversial change?

Comment (fredrikekre) [reply base/sysimg.jl]: See #22717
